crypto/common.go

```diff
@@ -25,6 +26,16 @@ const (
 	maxRelicPrgSeed = 1 << 32
 )
 
+func overwrite(data []byte) {
+	_, err := rand.Read(data) // checking err is enough
+	if err != nil {
+		// zero the buffer if randomizing failed
+		for i := 0; i < len(data); i++ {
+			data[i] = 0
+		}
+	}
+}
+
```

Comment:
is there any benefit in overwriting with random data instead of just writing zeros in the first place?

```suggestion
func overwrite(data []byte) {
	for i := range data {
		data[i] = 0
	}
}

```

BoringSSL simply memsets the memory on free:
```
void OPENSSL_cleanse(void *ptr, size_t len) {
#if defined(OPENSSL_WINDOWS)
  SecureZeroMemory(ptr, len);
#else
  OPENSSL_memset(ptr, 0, len);

#if !defined(OPENSSL_NO_ASM)
  /* As best as we can tell, this is sufficient to break any optimisations that
     might try to eliminate "superfluous" memsets. If there's an easy way to
     detect memset_s, it would be better to use that. */
  __asm__ __volatile__("" : : "r"(ptr) : "memory");
#endif
#endif  // !OPENSSL_NO_ASM
}
```

(with an additional caveat of preventing compiler optimizations.)

Reply:
Good question. First, this mitigation (either randomizing or zeroing) can only be helpful against "static" memory dump attacks that happen after the computation has ended. However, a more powerful attacker could use "dynamic" memory dumps during the computation to read sensitive data. In that case, the final "static" dump can help identify the memory areas to look at. Zeroing the sensitive areas gives an obvious hint for such attacks. 
Of course, an experimented attacker may avoid this mitigation by comparing multiple dynamic memory dumps and seeing the buffers evolution (especially that Golang initializes new allocated buffers with zeroes 😞 ). Randomizing instead of zeroing makes the attack slightly harder but not impossible. Randomizing is also a common practice in "grey box" cryptography implementations. 

Another limitation of overwriting buffers after use (either by randomizing or zeroing) is that it assumes the sensitive buffer was not copied in other areas. Such copies could have been made to the stack or registers, and these need to be overwritten as well. Looking at the assembly level for a specific compiler/architecture can answer the question, but that's not possible in our case (lib is written without a specific target compiler/arch). 
This mitigation is nice to have but I don't rely much on it, the library is still vulnerable to memory dumps and should only be run in safe environments.

Now back to the compiler optimizations: I am still wondering how to force the Go complier to not omit the call to `overwrite()`. Any thoughts about that?

Reply:
We may be overthinking it too much.  Our goal is to do a due diligence so that the memory we free won't have secret material in it.  As long as we can check that memset is not optimized out by compiler, we should be fine. (otherwise we down the rabbithole of software memory enclaves, mlock, PROT_NONE, and other "fun" system things)

For the long term solution, a `runtime/secret` is probably the right way of solving it: https://github.com/golang/go/issues/21865

PS. As far as physical access goes, I think this is an uphill battle: w/o enabled IOMMU one has full memory access via the PCIe DMA.

Reply:
> For the long term solution, a runtime/secret is probably the right way of solving it

Nice, yes I can wait for that to be available in Go. Thanks!

> As far as physical access goes, I think this is an uphill battle.

Once our model includes physical access, our library is totally insecure, even without memory access (because of side channel leakages). In my comment above, I was only considering remote memory access.  

My overall opinion:
- this mitigation makes the lib more secure only in a "weak" attack model. In other models it is not very relevant IMHO. 
- It doesn't cost much to add it, so I decided to add a slightly stronger mitigation, that is commonly used in grey box crypto implementations (overthinking there is required and common).
- It is noticeable in security audits so it's good to add it for that purpose.